R/data.table.R

```diff
@@ -716,7 +716,8 @@ replace_dot_alias = function(e) {
         # may evaluate to NULL | character() | "" | list(), likely a result of a user expression where no-grouping is one case being loop'd through
         bysubl = as.list.default(bysub)
         bysuborig = bysub
-        if (is.name(bysub) && !(bysub %chin% names_x)) {  # TO DO: names(x),names(i),and i. and x. prefixes
+        # as.character here as a workaround for #4708
+        if (is.name(bysub) && !(as.character(bysub) %chin% names_x)) {  # TO DO: names(x),names(i),and i. and x. prefixes
```

Comment:
An extra comment on this line to remind us in future? `# as.character() is for #4708`

Reply:
in addition to the line above?

Reply:
sorry, I don't follow. What line is the line above? This line (743 of data.table.R) has a prior comment `# TO DO: names(x),names(i),and i. and x. prefixes`.  You've added the `as.character()` around `bysub` to fix this non-ascii bug and it's not clear to me why doing that (i.e. adding `as.character()`) fixes this, but I don't need to know why it works, happy to accept it as the test proves it works. But when I look in future at this line, I know I'll be wondering why `as.character()` is there because I know that `%chin%` accepts symbols without needing `as.character()`, so I was suggesting to add `as.character()` is for #4708` as an extra comment on this line.  But ... now having written this long comment ... makes me wonder if the `as.character()` should actually be moved to be inside `%chin%` instead to fix any other calls to `%chin%` that we don't know yet have the same problem with non-ascii characters in `name`s; i.e. just add `x = as.character(x)` as the first line to `%chin%` with a comment `# as.character() is for #4708`.

Reply:
that indeed sounds better as long as it's `if (!is.character(x)) x<-as.character(x)`.

Reply:
I'm 95% sure that `as.character()` has a switch inside it already that returns its input straight back (no copy) if the input is already `character`. We rely on that principle quite a lot in data.table R code (at least I do) so if that's not true we have a sweep to do to add a bunch of `if()`s like this. Note that `as.character` is a primitive so there isn't any dispatch overhead.
Did you have something else in mind?

Reply:
Hmm I think you're right. I had in mind that we're careful to make sure `as.data.table()` always copies. But seems there's no copy:

```
x <- "a"
address(x)
# [1] "0x557a49d38778"
address(as.character(x))
# [1] "0x557a49d38778"
y <- x
address(y)
# [1] "0x557a49d38778"
address(as.character(y))
# [1] "0x557a49d38778"
```